core/src/avm2/globals/flash/display/bitmapdata.rs

```diff
@@ -375,6 +375,330 @@ pub fn get_pixel32<'gc>(
+        let x = args
+            .get(0)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_u32(activation)?;
+        let y = args
+            .get(1)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_u32(activation)?;
+        let color = args
+            .get(2)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_i32(activation)?;
+        bitmap_data
+            .write(activation.context.gc_context)
+            .set_pixel(x, y, color.into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Implements `BitmapData.setPixel32`.
+pub fn set_pixel32<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(bitmap_data) = this.and_then(|t| t.as_bitmap_data()) {
+        let x = args
+            .get(0)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_i32(activation)?;
+        let y = args
+            .get(1)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_i32(activation)?;
+        let color = args
+            .get(2)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_i32(activation)?;
+        bitmap_data
+            .write(activation.context.gc_context)
+            .set_pixel32(x, y, color.into());
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Implements `BitmapData.setPixels`.
+pub fn set_pixels<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let rectangle = args
+        .get(0)
+        .unwrap_or(&Value::Undefined)
+        .coerce_to_object(activation)?;
+
+    let bytearray = args
+        .get(1)
+        .unwrap_or(&Value::Undefined)
+        .coerce_to_object(activation)?;
+    if let Some(bitmap_data) = this.and_then(|t| t.as_bitmap_data()) {
+        let x = rectangle
+            .get_property(&Multiname::public("x"), activation)?
+            .coerce_to_u32(activation)?;
+        let y = rectangle
+            .get_property(&Multiname::public("y"), activation)?
+            .coerce_to_u32(activation)?;
+        let width = rectangle
+            .get_property(&Multiname::public("width"), activation)?
+            .coerce_to_u32(activation)?;
+        let height = rectangle
+            .get_property(&Multiname::public("height"), activation)?
+            .coerce_to_u32(activation)?;
+
+        let ba_read = bytearray
+            .as_bytearray()
+            .ok_or("ArgumentError: Parameter must be a bytearray")?;
+
+        let mut bitmap_data = bitmap_data.write(activation.context.gc_context);
+        let mut ind = 0;
+
+        for y in y..y + height {
+            for x in x..x + width {
+                // Copy data from bytearray until EOFError or finished
+                if let Ok(color) = ba_read.read_int_at(ind) {
+                    bitmap_data.set_pixel32_raw(x, y, color.into());
```

Comment:
I think this should be `set_pixel32, which will correctly discard the transparency information if `bitmap_data` is not opaque. Can you add a test for this?

Reply:
Fixed. It also made me notice I should have advanced index by 4.